hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/viewfs/ViewFileSystem.java

```diff
@@ -118,38 +118,22 @@ protected FsGetter fsGetter() {
    * Caching children filesystems. HADOOP-15565.
    */
   static class InnerCache {
-    private Map<Key, FileSystem> map = new HashMap<>();
+    private ConcurrentHashMap<Key, FileSystem> map = new ConcurrentHashMap<>();
     private FsGetter fsCreator;
-    private ReentrantReadWriteLock rwLock = new ReentrantReadWriteLock();
 
     InnerCache(FsGetter fsCreator) {
       this.fsCreator = fsCreator;
     }
 
     FileSystem get(URI uri, Configuration config) throws IOException {
       Key key = new Key(uri);
-      FileSystem fs = null;
-      try {
-        rwLock.readLock().lock();
-        fs = map.get(key);
-        if (fs != null) {
-          return fs;
-        }
-      } finally {
-        rwLock.readLock().unlock();
-      }
-      try {
-        rwLock.writeLock().lock();
-        fs = map.get(key);
-        if (fs != null) {
-          return fs;
-        }
-        fs = fsCreator.getNewInstance(uri, config);
-        map.put(key, fs);
-        return fs;
-      } finally {
-        rwLock.writeLock().unlock();
+      if (map.contains(key)) {
+        return map.get(key);
       }
+
+      FileSystem fs = fsCreator.getNewInstance(uri, config);
```

Comment:
* this does not have the same concurrency semantics as the existing code
* fsCreator.getNewInstance should only happen once in existing code with lock
* in new code, concurrent threads could call fsCreator.getNewInstance multiple times
* once the putIfAbsent is called, then future threads should no longer call fsCreator.getNewInstance
* the new behaviour may be acceptable - but it is important to analyse if it is ok to allow multiple fsCreator.getNewInstance calls
* do you have evidence to show that there is a performance bottleneck with the existing code?
* one option is to keep the lock but use a concurrent map nonetheless - this code would be similar to your change - the main diff is

```
 private ConcurrentHashMap<Key, FileSystem> map = new ConcurrentHashMap<>();
 private ReentrantLock lock = new ReentrantLock();

if (map.contains(key)) {
   return map.get(key);
}
try {
  lock.lock();
  // need to check if another thread has put an fs for this key while we waited for the lock
  FileSystem fs = map.get(key);
  if (fs == null) {
    fs = fsCreator.getNewInstance(uri, config);
    map.put(key, fs);
  }
  return fs;
} finally {
  lock.unlock();
}
```

This avoids the read locks and the write lock only happens when the key has no fs stored for it.

Reply:
I would also suggest looking at computeIfAbsent instead of putIfAbsent.

https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#computeIfAbsent-K-java.util.function.Function-

With computeIfAbsent, you may be able to avoid the lock but still have the behaviour that only one fs is created per key.

Reply:
Hi @pjfanning, 

Thanks for your comments. Replaced with computeIfAbsent(). Could you take another look?